### ISSUE TITLE
Support for JSON log events logged to STDOUT

### ISSUE THREAD

**Issue by legdba** [kubernetes/kubernetes#15795]
JSON log events logged to STDOUT are handled has a raw string wrapped in a new JSON document with additional meta-date. This cause the application initial JSON document to be mostly useless.

Instead JSON log events logged to STDOUT should be handled as JSON documents preserved and merged with additional meta-data.

**Comment by jlowdermilk** [comment]
Sorry can you add more context/clarification? What do you mean by JSON log events? Which logs are you referring to, container logs?

**Comment by legdba** [comment]
*(note: this has initially been discussed here: https://groups.google.com/forum/#!topic/google-containers/ZD19Hy8qIlo)*

When an application within a container logs events as one-line JSON documents to STDOUT, this JSON document is used as a string, not as a JSON document. This means the application JSON document attributes cannot be used in ElasticSearch/Kibana. For me K8S should be friendly with applications logging in JSON to STDOUT since this is now a best practice.

Let me give you a working example for clarity and to allow anybody to test.

I have a test Java REST server ([available here](https://github.com/legdba/servicebox-jaxrs)) logging in JSON to STDOUT and it obviously runs within a Docker container ([available here](https://quay.io/repository/legdba/servicebox-jaxrs))

```shell
$ docker run -ti -p :8080:8080 --rm=true quay.io/legdba/servicebox-jaxrs:latest
java -XX:OnOutOfMemoryError=kill -9 %p -Xms50m -Xmx50m -cp /opt/servicebox-jaxrs-r30-f50d076-master/lib/* com.brimarx.servicebox.EmbededServer -p 8080
{"@timestamp":"2015-10-21T07:02:42.508+00:00","@version":1,"message":"logs initialized","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"INFO","level_value":20000,"HOSTNAME":"3e5be690cb9f"}
{"@timestamp":"2015-10-21T07:02:43.462+00:00","@version":1,"message":"o.e.j.s.h.ContextHandler@5c90e579{/,null,null} contextPath ends with /","logger_name":"org.eclipse.jetty.server.handler.ContextHandler","thread_name":"main","level":"WARN","level_value":30000,"HOSTNAME":"3e5be690cb9f"}
{"@timestamp":"2015-10-21T07:02:43.621+00:00","@version":1,"message":"server starting...","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"INFO","level_value":20000,"HOSTNAME":"3e5be690cb9f"}
{"@timestamp":"2015-10-21T07:02:46.227+00:00","@version":1,"message":"server started; serving requests on port 8080 ...","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"WARN","level_value":30000,"HOSTNAME":"3e5be690cb9f"}
```

This app outputs a 1st raw-string log event (for testing, to check how raw strings are managed; Java can output to STDOUT for GC logs and crash logs -> we don't want to lose those raw-strings) and then one-line JSON documents per log event.

Then I'm setting up a GKE K8S cluster:
```shell
export KUBE_LOGGING_DESTINATION=elasticsearch
export KUBE_ENABLE_NODE_LOGGING=true
export KUBERNETES_PROVIDER=aws
curl -sS https://get.k8s.io | bash
```

When the K8S cluster is setup I start my app:

file: servicebox-jaxrs-r53-ca9916d-master.rc.yaml
```yaml
apiVersion: v1
kind: ReplicationController
metadata:
  name: servicebox-jaxrs-r53-ca9916d-master
spec:
  # Number of running replicas; defaults to 1
  replicas: 1
  # selector identifies the set of Pods that this
  # replication controller is responsible for managing
  selector:
    name: servicebox
  # podTemplate defines the 'cookie cutter' used for creating
  # new pods when necessary
  template:
    metadata:
      labels:
        # Important: these labels need to match the selector above
        # The api server enforces this constraint.
        name:       servicebox         # application name
        ver:        r53-ca9916d-master # application version
        tier:       frontend           # network tier: frontend | backend
        env:        staging            # deployment environment: dev | qa | staging | production
        role:       active             # active | master | slave | primary | secondary
        track:      canary             # canary | stable
        provider:   legdba             # provider name
    spec:
      containers:
      - name: servicebox
        image: quay.io/legdba/servicebox-jaxrs:r53-ca9916d-master
        args: ['--port', '$(PORT)']
        ports:
        - containerPort: 80
          name: http
        env:
          - name: PORT
            value: "80"
```

Kubectl command:
```shell
kubectl create -f servicebox-jaxrs-r53-ca9916d-master.rc.yaml
```

The associated container logs are:
```shell
kubectl logs servicebox-jaxrs-r53-ca9916d-master-9it6n
java -XX:OnOutOfMemoryError=kill -9 %p -Xms50m -Xmx50m -cp /opt/servicebox-jaxrs-r53-ca9916d-master/lib/* com.brimarx.servicebox.EmbededServer --port 80
{"@timestamp":"2015-10-21T08:53:38.736+00:00","@version":1,"message":"logs initialized","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"INFO","level_value":20000,"HOSTNAME":"servicebox-jaxrs-r53-ca9916d-master-9it6n"}
{"@timestamp":"2015-10-21T08:53:40.116+00:00","@version":1,"message":"o.e.j.s.h.ContextHandler@55f616cf{/,null,null} contextPath ends with /","logger_name":"org.eclipse.jetty.server.handler.ContextHandler","thread_name":"main","level":"WARN","level_value":30000,"HOSTNAME":"servicebox-jaxrs-r53-ca9916d-master-9it6n"}
{"@timestamp":"2015-10-21T08:53:40.323+00:00","@version":1,"message":"server starting...","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"INFO","level_value":20000,"HOSTNAME":"servicebox-jaxrs-r53-ca9916d-master-9it6n"}
{"@timestamp":"2015-10-21T08:53:42.819+00:00","@version":1,"message":"server started; serving requests on port 80 ...","logger_name":"com.brimarx.servicebox.EmbededServer","thread_name":"main","level":"WARN","level_value":30000,"HOSTNAME":"servicebox-jaxrs-r53-ca9916d-master-9it6n"}
```

Searching in Kibana for the associated logs the following entries are returned:
```json
{
  "_index": "logstash-2015.10.21",
  "_type": "fluentd",
  "_id": "AVCJmwsg8AIzg1-0C0yM",
  "_score": null,
  "_source": {
    "log": "java -XX:OnOutOfMemoryError=kill -9 %p -Xms50m -Xmx50m -cp /opt/servicebox-jaxrs-r53-ca9916d-master/lib/* com.brimarx.servicebox.EmbededServer --port 80\n",
    "stream": "stdout",
    "tag": "kubernetes.servicebox-jaxrs-r53-ca9916d-master-9it6n_default_servicebox",
    "@timestamp": "2015-10-21T08:53:36+00:00"
  },
  "fields": {
    "@timestamp": [
      1445417616000
    ]
  },
  "highlight": {
    "log": [
      "@kibana-highlighted-field@java@/kibana-highlighted-field@ -@kibana-highlighted-field@XX:OnOutOfMemoryError@/kibana-highlighted-field@=@kibana-highlighted-field@kill@/kibana-highlighted-field@ -@kibana-highlighted-field@9@/kibana-highlighted-field@ %@kibana-highlighted-field@p@/kibana-highlighted-field@ -@kibana-highlighted-field@Xms50m@/kibana-highlighted-field@ -@kibana-highlighted-field@Xmx50m@/kibana-highlighted-field@ -cp /opt/servicebox-jaxrs-r53-ca9916d-master"
    ]
  },
  "sort": [
    1445417616000
  ]
}
{
  "_index": "logstash-2015.10.21",
  "_type": "fluentd",
  "_id": "AVCJmwsg8AIzg1-0C0yQ",
  "_score": null,
  "_source": {
    "log": "{\"@timestamp\":\"2015-10-21T08:53:42.819+00:00\",\"@version\":1,\"message\":\"server started; serving requests on port 80 ...\",\"logger_name\":\"com.brimarx.servicebox.EmbededServer\",\"thread_name\":\"main\",\"level\":\"WARN\",\"level_value\":30000,\"HOSTNAME\":\"servicebox-jaxrs-r53-ca9916d-master-9it6n\"}\n",
    "stream": "stdout",
    "tag": "kubernetes.servicebox-jaxrs-r53-ca9916d-master-9it6n_default_servicebox",
    "@timestamp": "2015-10-21T08:53:42+00:00"
  },
  "fields": {
    "@timestamp": [
      1445417622000
    ]
  },
  "highlight": {
    "log": [
      "\":\"WARN\",\"level_value\":30000,\"HOSTNAME\":\"@kibana-highlighted-field@servicebox@/kibana-highlighted-field@-@kibana-highlighted-field@jaxrs@/kibana-highlighted-field@-@kibana-highlighted-field@r53@/kibana-highlighted-field@-@kibana-highlighted-field@ca9916d@/kibana-highlighted-field@-@kibana-highlighted-field@master@/kibana-highlighted-field@-@kibana-highlighted-field@9it6n@/kibana-highlighted-field@\"}\n",
      " requests on port 80 ...\",\"logger_name\":\"@kibana-highlighted-field@com.brimarx.servicebox.EmbededServer@/kibana-highlighted-field@\",\"thread_name\":\"main\",\"level"
    ]
  },
  "sort": [
    1445417622000
  ]
}
{
  "_index": "logstash-2015.10.21",
  "_type": "fluentd",
  "_id": "AVCJmwsg8AIzg1-0C0yP",
  "_score": null,
  "_source": {
    "log": "{\"@timestamp\":\"2015-10-21T08:53:40.323+00:00\",\"@version\":1,\"message\":\"server starting...\",\"logger_name\":\"com.brimarx.servicebox.EmbededServer\",\"thread_name\":\"main\",\"level\":\"INFO\",\"level_value\":20000,\"HOSTNAME\":\"servicebox-jaxrs-r53-ca9916d-master-9it6n\"}\n",
    "stream": "stdout",
    "tag": "kubernetes.servicebox-jaxrs-r53-ca9916d-master-9it6n_default_servicebox",
    "@timestamp": "2015-10-21T08:53:40+00:00"
  },
  "fields": {
    "@timestamp": [
      1445417620000
    ]
  },
  "highlight": {
    "log": [
      "...\",\"logger_name\":\"@kibana-highlighted-field@com.brimarx.servicebox.EmbededServer@/kibana-highlighted-field@\",\"thread_name\":\"main\",\"level\":\"INFO\",\"level_value\":20000,\"HOSTNAME\":\"@kibana-highlighted-field@servicebox@/kibana-highlighted-field@-@kibana-highlighted-field@jaxrs@/kibana-highlighted-field@-@kibana-highlighted-field@r53@/kibana-highlighted-field@-@kibana-highlighted-field@ca9916d@/kibana-highlighted-field@-@kibana-highlighted-field@master@/kibana-highlighted-field@-@kibana-highlighted-field@9it6n@/kibana-highlighted-field@\"}\n"
    ]
  },
  "sort": [
    1445417620000
  ]
}
...
```

The 1st event is handled as raw-string and that is ok.
But following events which are JSON documents are handled as raw string as well and the application JSON attributes such as *logger_name*, *level* or *thread-name* cannot be used for indexing, searches and so on.

Instead, the Elastic Search entries should IMHO look like this:
```json
... (1st event omited) ...
{
  "_index": "logstash-2015.10.21",
  "_type": "fluentd",
  "_id": "AVCJmwsg8AIzg1-0C0yQ",
  "_score": null,
  "_source": {
  "@version":1,
  "logger_name":"com.brimarx.servicebox.EmbededServer",
  "thread_name":"main",
  "level":"WARN",
  "level_value":30000,
  "HOSTNAME":"servicebox-jaxrs-r53-ca9916d-master-9it6n",
  "log": "server started; serving requests on port 80 ...",
  "stream": "stdout",
  "tag": "kubernetes.servicebox-jaxrs-r53-ca9916d-master-9it6n_default_servicebox",
  "@timestamp": "2015-10-21T08:53:42+00:00"
  },
  "fields": {
    "@timestamp": [
      1445417622000
    ]
  },
...
```

In this example all application JSON attributes are merged in the ElasticSearch JSON document and can hence be used for searching and indexing. Maybe the example structure is not the right one (attributes could clash as an example) but this is the idea.

I hope this is clear :)

**Comment by jlowdermilk** [comment]
Thanks for the added detail. Docker log parsing rules are defined in this [td-agent.conf](https://github.com/kubernetes/kubernetes/blob/b9cfab87e33ea649bdd13a1bd243c502d76e5d22/cluster/addons/fluentd-elasticsearch/fluentd-es-image/td-agent.conf). If you see an obvious fix feel free to send a pr. cc @dchen1107

**Comment by legdba** [comment]
AFAIK the Fluentd standard plugins do not allow to fix this with configuration only. Some Fluentd new plugin would be required (https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter could be a good starting point). I'll think about it and check what exists already.

**Comment by legdba** [comment]
K8S is using https://github.com/fabric8io/fluent-plugin-kubernetes_metadata_filter already which support JSON merging. Looking at it it looks like adding a "merge_json_log true" would do the trick. I'll work on a patch and PR then. I'm not sure how to add automated tests for this though, any guidance on this?

**Comment by jlowdermilk** [comment]
There is [an e2e test](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/es_cluster_logging.go) for cluster logging with elastic search. You could adapt the [logger pods](https://github.com/kubernetes/kubernetes/blob/master/test/e2e/es_cluster_logging.go#L212) created to output json, and verify the log lines are correctly parsed.

**Comment by legdba** [comment]
Fixed in Kubernetes 1.1.
